common/src/main/java/org/apache/celeborn/common/util/PushDataHeaderUtils.java

```diff
@@ -0,0 +1,104 @@
+ */
+
+package org.apache.celeborn.common.util;
+
+import java.util.zip.CRC32;
+
+import org.apache.celeborn.common.unsafe.Platform;
+
+public class PushDataHeaderUtils {
+  // Data Header Layout:
+  // | mapId (4 bytes)                        |
+  // | attemptId (4 bytes)                    |
+  // | batchId with checksum flag (4 bytes)   |
+  // | length with checksum length (4 bytes)  |
+  // | checksum (4 bytes)                     |
+  //
+  // Fields description:
+  // - mapId: Unique identifier for the map (4 bytes)
+  // - attemptId: Identifier for the attempt (4 bytes)
+  // - batchId with checksum flag:
+  //   -- checksum flag: 1 bit (indicates if batchId has a checksum)
+  //   -- batchId: 31 bits (always positive when represented as an integer)
+  // - length with checksum length: total length of the data + 4 bytes for checksum
+  // - checksum: Always positive integer (4 bytes)
+  public static final int BATCH_HEADER_SIZE = 5 * 4;
+  public static final int BATCH_HEADER_SIZE_WITHOUT_CHECKSUM = BATCH_HEADER_SIZE - 4;
+  public static final int MAP_ID_OFFSET = Platform.BYTE_ARRAY_OFFSET;
+  public static final int ATTEMPT_ID_OFFSET = Platform.BYTE_ARRAY_OFFSET + 4;
+  public static final int BATCH_ID_OFFSET = Platform.BYTE_ARRAY_OFFSET + 8;
+  public static final int LENGTH_OFFSET = Platform.BYTE_ARRAY_OFFSET + 12;
+  public static final int CHECKSUM_OFFSET = Platform.BYTE_ARRAY_OFFSET + 16;
+  public static final int POSITIVE_MASK = 0x7FFFFFFF;
+  public static final int HIGHEST_1_BIT_FLAG_MASK = 0x80000000;
+
+  public static void buildDataHeader(
+      byte[] data, int mapId, int attemptId, int batchId, int length) {
+    assert data.length >= BATCH_HEADER_SIZE;
+    int batchIdWithChecksumFlag = batchIdWithChecksumFlag(batchId);
+    int lengthWithChecksum = length + 4;
+    Platform.putInt(data, MAP_ID_OFFSET, mapId);
+    Platform.putInt(data, ATTEMPT_ID_OFFSET, attemptId);
+    Platform.putInt(data, BATCH_ID_OFFSET, batchIdWithChecksumFlag);
+    Platform.putInt(data, LENGTH_OFFSET, lengthWithChecksum);
+    Platform.putInt(data, CHECKSUM_OFFSET, computeHeaderChecksum32(data));
+  }
+
+  public static int batchIdWithChecksumFlag(int batchId) {
+    return batchId | HIGHEST_1_BIT_FLAG_MASK;
+  }
+
+  public static int batchIdWithoutChecksumFlag(int batchId) {
+    return batchId & POSITIVE_MASK;
+  }
+
+  public static boolean hasChecksumFlag(byte[] data) {
+    int batchId = Platform.getInt(data, BATCH_ID_OFFSET);
+    return (batchId & HIGHEST_1_BIT_FLAG_MASK) != 0;
+  }
+
+  public static int getMapId(byte[] data) {
+    return Platform.getInt(data, MAP_ID_OFFSET);
+  }
+
+  public static int getAttemptId(byte[] data) {
+    return Platform.getInt(data, ATTEMPT_ID_OFFSET);
+  }
+
+  public static int getBatchId(byte[] data) {
+    return batchIdWithoutChecksumFlag(Platform.getInt(data, BATCH_ID_OFFSET));
+  }
+
+  public static int getLength(byte[] data) {
+    return Platform.getInt(data, LENGTH_OFFSET) - 4;
```

Comment:
This will cause errors if an old client reads from Celeborn workers with this feature.

Reply:
+1, we should think about how maintain compatibility with older versions.